kubeadm: implementation of `UpgradeConfiguration` API types

chendave opened the pull request:
Signed-off-by: Dave Chen <dave.chen@arm.com>

/kind feature


<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of #https://github.com/kubernetes/kubeadm/issues/2489
Part of #https://github.com/kubernetes/kubeadm/issues/2890

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
https://docs.google.com/document/d/1R4zLtgDadNM8_N2VzS0DI2mADcYmkyJ3L7lJVpbyZsA
```


TODO:
* [x] migrated to v1beta4 when the API version is ready
* [x] split the cleanup out from this PR or in separate  commits
* [x] fetch configmaps from cluster always: https://github.com/kubernetes/kubeadm/issues/2785

chendave commented:
/hold

-  going to be migrate to v1beta4 when the API version is ready.
- will split into couple of commits as this pr contains several cleanup as well.

chendave commented:
Just took some time to rebase, and also cc to all active contributor to level up the visibility. @pacoxu @SataQiu @neolit123  

Basically, the change to support resetCfg (https://github.com/kubernetes/kubernetes/pull/113583) and the upgradeCfg (https://github.com/kubernetes/kubernetes/pull/114062) is ready.

I understood we still need to wait for new API ready or well prepared in the advance,  I would appreciated for any early review / comments for those two change.

pacoxu commented:
If I understand correctly, the comments in https://docs.google.com/document/d/1R4zLtgDadNM8_N2VzS0DI2mADcYmkyJ3L7lJVpbyZsA/edit# shows that we should add new API in new API version v1beta4.

chendave commented:
> If I understand correctly, the comments in https://docs.google.com/document/d/1R4zLtgDadNM8_N2VzS0DI2mADcYmkyJ3L7lJVpbyZsA/edit# shows that we should add new API in new API version v1beta4.

yes, migrate to the new API should be easy.

pacoxu commented:
> yes, migrate to the new API should be easy.

Do you mean that we can start reviewing this PR now or we should wait until this is migrated to the new API?
Is this targeted for v1.27 or the next release cycle(maybe v1.28)?

chendave commented:
@pacoxu pls help to review if you have some bandwidth or else it is fine to defer this to a future release cycle. AFAIK, this is not targeted for any version yet.

SataQiu commented:
> If I understand correctly, the comments in https://docs.google.com/document/d/1R4zLtgDadNM8_N2VzS0DI2mADcYmkyJ3L7lJVpbyZsA/edit# shows that we should add new API in new API version v1beta4.

Yes, v1beta3 is locked and new features cannot be added. Maybe we should first update KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/970-kubeadm-config

chendave committed:
kubeadm: Add `UpgradeConfiguration` for v1beta4

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave committed:
kubeadm: fix fuzzing for upgradeConfiguration

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave committed:
kubeadm: Validate mixed arguments for `kubeadm upgrade plan/apply/diff/node`

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented:
/test pull-kubernetes-integration

check is that something cause by the pr (kubeadm has the builddata set to empty string)

>    version_test.go:72: 'kubeadm version ' stdout did not match expected regex; wanted: ["^kubeadm version: &version\\.Info{Major:\".+\", Minor:\".+\", GitVersion:\".+\", GitCommit:\".+\", GitTreeState:\".+\", BuildDate:\".+\", GoVersion:\".+\", Compiler:\".+\", Platform:\".+\"}\n$"], got: [kubeadm version: &version.Info{Major:"1", Minor:"28+", GitVersion:"v1.28.0-alpha.4.440+def9d7059bc688", GitCommit:"def9d7059bc6883033a18e880f2bb365b6d1470e", GitTreeState:"clean", BuildDate:"", GoVersion:"go1.20.6", Compiler:"gc", Platform:"linux/amd64"}

chendave commented:
@neolit123 @SataQiu @pacoxu this is ready for review! pls let me know whether I'm on the right track, thanks!

chendave commented:
kindly ping @neolit123 @SataQiu @pacoxu for review, 1.29 release cycle is started, we need to speed up a little bit for v1beta4, we still have lots of stuff left for v1beta4.

chendave commented:
/hold cancel

/assign @SataQiu @neolit123

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -95,12 +95,10 @@ func loadConfig(cfgPath string, kubeConfigPath string, skipComponentConfigs bool
-	// missing from the file will be automatically regenerated by kubeadm even if they are present in the cluster.
-	// The resulting configs overwrite the existing cluster ones at the end of a successful upgrade apply operation.
+	// If there are kubeadm types (API group kubeadm.kubernetes.io) present, we need to keep the existing behavior here.
+	// Component configs that are missing from the file will be automatically regenerated by kubeadm if they are not present in the cluster.
```
oh, i forgot about this part.
currently, users can pass a --config to kubeadm upgrade that has the componentconfigs for kubelet and kube-proxy present, separated by ---.

this was done to allow component config upgrades, as there is no other UX for the user.
i was not a big fan of this when it was done, as it steps over the mutation on upgrade boundary (not desired).

IMO, upgrade's --config should not allow mixing UpgradeConfiguration with the component config types.
WDYT?

neolit123 replied:
more thoughts, by not allowing InitConfigiration on upgrade --config any more, we are effectively breaking users that pass the component config on upgrade.

perhaps we should not allow the component config on upgrade and that's all.
this might need some more discussion.

chendave replied:
> --config should not allow mixing UpgradeConfiguration with the component config types.

I don't have enough context on this, I am not sure is there any user are using the mixing configuraition. @SataQiu @pacoxu  thoughts?

> by not allowing InitConfigiration on upgrade --config any more, we are effectively breaking users that pass the component config +init configuration on upgrade.

- we should deprecate the use of `InitConfigiration` on upgrade, as there are some upgrade specific configuration which `InitConfigiration` cannot provide.
- Have a `InitConfigiration` in the upgrade data, built it from the current cluster and is immutable, while user's input of a Init config file means the upgrade might mutable via the new configuration.

chendave replied:
the whole logic is refactored.

neolit123 reviewed (commented):
@chendave thank you for working on this difficult addition of UpgradeConfiguration.
i did my best on a first pass.

i will kindly ask you to be patient with this change...it might take a while.
we are facing some complicated issues at hand, such as:
- what to embed in UpgradeConfiguration.
- should we allow component configs
- what refactor to do in this PR.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+type UpgradeConfiguration struct {
+	metav1.TypeMeta
+
+	// ApiServerManifestPath is the path to API server manifest
+	// Default: /etc/kubernetes/manifests/kube-apiserver.yaml
+	ApiServerManifestPath string
```
Just a question? Do we need to consider the case of dry-run? 

Looks like to me `kubeadm upgrade diff` (only this CLI needs this flag I think)  is not created for the comparison with dry-run manifest.

https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/cmd/kubeadm/app/cmd/init.go#L429-L431

neolit123 replied:
i still haven't checked the source code about where these custom manifest paths are used (currently on phone).

if the use case is important this could have been a custom folder with standard manifest file names. not custom manifest path for each component.

IMO, we should deprecate the flags and not add these options in config.

chendave replied:
update: I paste the wrong link, it should be, 

https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/cmd/kubeadm/app/cmd/init.go#L457-L462

So, it will use the hard-coded path to write the manifest, the same for join and reset, but if it's dry-run mode, the path can be a little different and can be set by the env `KUBEADM_INIT_DRYRUN_DIR`. 

But I think the dir is just a temp dir and my guess for the `kubeadm upgrade diff` is not created for `dry-run`, I am fine to remove it from config file.

@SataQiu @pacoxu  any inputs for this?

chendave replied:
I removed them from the upgrade config API type.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	// Default: /etc/kubernetes/manifests/kube-apiserver.yaml
+	ApiServerManifestPath string
+
+	// AllowExperimentalUpgrades instruct kubeadm to show unstable versions of Kubernetes as an upgrade
```
> does not allow code suggestions (+ button) for some reason.

me too 😄 , you can apply the suggestion manually with markdown syntax, e.g.

```suggestion
 // AllowExperimentalUpgrades instructs kubeadm to show unstable versions of Kubernetes as an upgrade
```

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	CertificateRenewal *bool
+
+	// ClusterConfiguration holds the cluster-wide information, and embeds that struct (which can be (un)marshalled separately as well)
+	ClusterConfiguration `json:"-"`
```
> is it too complex to just pass upgrade configuration separately from cluster configuration around separately instead of this embedding?

I will take a look and get back to you later on this.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	ClusterConfiguration `json:"-"`
+
+	// ContextLines is the number of lines of context in the diff
+	ContextLines int
```
> is the flag just --context-lines?

yes, it's only used for `kubeadm upgrade diff`

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+
+	// ControllerManagerManifestPath is the path to the API server manifest
+	// Default: /etc/kubernetes/manifests/kube-controller-manager.yaml
+	ControllerManagerManifestPath string
```
will remove this if we agree we won't consider the case of dry-run.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	// DryRun tells if the dry run mode is enabled, don't apply any change if it is and just output what would be done.
+	DryRun bool
+
+	// NodeRegistration holds the node specific information needed for upgrade, e.g. CRISocket, kubeadm needs this info in the phase of preflight to pull the required images
```
Done

neolit123 commented:
one Q we have to ask is do we want to move diff specific options in a substruct?

e.g. UpgradeConfiguration.Diff.SomeOptionSpecificToDiff

currently the KubeletConfiguration v1beta1 has many options mixed in the same config level and it's a bit messy.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	NodeRegistration NodeRegistrationOptions
+
+	// EtcdUpgrade instructs kubeadm to execute etcd upgrade during upgrades
+	// Default: true
```
removed for this version.

chendave replied:
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -29,6 +29,7 @@ limitations under the License.
 //     Use `APIServer.ExtraEnvs`, `ControllerManager.ExtraEnvs`, `Scheduler.ExtraEnvs`,  `Etcd.Local.ExtraEnvs`.
 //   - The ResetConfiguration API type is now supported in v1beta4. Users are able to reset a node by passing a --config file to "kubeadm reset".
 //   - `dry-run` mode in is now configurable in InitConfiguration and JoinConfiguration config files.
+//   - The UpgradeConfiguration API type is now supported in v1beta4. Users are able to upgrade a cluster or node by passing a --config file to each sub-command of "kubeadm upgrade".
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	SkipPhases []string
+
+	// UpgradeManifestTimeout is timeout of upgrading the static pod manifest.
+	UpgradeManifestTimeout *metav1.Duration
```
it is used here and hardcoded as 5mins, https://github.com/kubernetes/kubernetes/blob/4f874a224a31a5dc5745a0c140d3769bb178f811/cmd/kubeadm/app/cmd/phases/upgrade/node/controlplane.go#L76

make it configurable via the config file in the change.

neolit123 replied:
if you search in k/kubeadm for api changes there is one for controlling various timeouts. it"s not clear to me how we want to structure the whole thing and whether adding this tomeout in upgradeconfiguration the way it is in this PR is the right thing to do. users did not request it  so maybe we shouldn't add it.

chendave replied:
> if you search in k/kubeadm for api changes there is one for controlling various timeouts.

we need to take care of this one for v1beta4 too.

> users did not request it so maybe we shouldn't add it.

will remove it from the config file to reduce the complexity.

chendave replied:
removed.

if there is any -1 from other reviewers, I will add it back.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	// Default: true
+	EtcdUpgrade *bool
+
+	// ForceUpgrade flag instruct kubeadm to upgrade the cluster without prompting for confirmation
```
Done

chendave commented on cmd/kubeadm/app/constants/constants.go:
```diff
@@ -224,6 +224,9 @@ const (
 	DefaultControlPlaneTimeout = 4 * time.Minute
 
+	// UpgradeConfigurationKind is the string kind value for the UpgradeConfiguration struct
+	UpgradeConfigurationKind = "UpgradeConfiguration"
```
Done

chendave commented on cmd/kubeadm/app/util/config/cluster.go:
```diff
@@ -256,3 +257,83 @@ func getRawAPIEndpointFromPodAnnotationWithoutRetry(client clientset.Interface,
 }
+
+// TODO: De-dup the code here as most of the code is similar with FetchInitConfigurationFromCluster.
+// FetchUpgradeConfigurationFromCluster fetches configuration from a ConfigMap in the cluster
```
Done

chendave replied:
this logic is removed, no TODO left for this I think.

chendave commented on cmd/kubeadm/app/util/config/upgradeconfiguration.go:
```diff
@@ -0,0 +1,147 @@
+
+// documentMapToUpgradeConfiguration takes a map between GVKs and YAML documents (as returned by SplitYAMLDocuments),
+// finds a UpgradeConfiguration, decodes it, dynamically defaults it and then validates it prior to return.
+func documentMapToUpgradeConfiguration(gvkmap kubeadmapi.DocumentMap, allowDeprecated bool) (*kubeadmapi.UpgradeConfiguration, error) {
```
We can leave this as a follow up review as reset does, this pr is already too big.

neolit123 replied:
please mention all followup PRs as TODO comments in the tracking issue, so that we don't forget.

chendave replied:
updated https://github.com/kubernetes/kubeadm/issues/2890 to have a work item there.

chendave commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -95,6 +98,12 @@ func loadConfig(cfgPath string, client clientset.Interface, skipComponentConfigs
 		return cfg, true, err
 	}
 
+	// We are done here!
+	// Whatever it has in the config file, there is no way for us to build the UpgradeConfiguration without a valid client.
+	if client == nil {
+		return nil, false, errors.New("no valid client, cannot fetch upgrade configuration from cluster")
+	}
+
```
I am afraid not, this code branch is handle with the config file, it's belong to different code path.

neolit123 replied:
oh right, my mistake.

neolit123 commented on cmd/kubeadm/app/util/marshal_test.go:
```diff
@@ -437,3 +437,38 @@ func TestGroupVersionKindsHasResetConfiguration(t *testing.T) {
+
+	for _, rt := range tests {
+		t.Run(rt.name, func(t2 *testing.T) {
+
```
```suggestion
```

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	// Default: /etc/kubernetes/manifests/kube-scheduler.yaml
+	SchedulerManifestPath string
+
+	// SkipPhases is a list of phases to skip during command execution.
+	// The list of phases can be obtained with the "kubeadm upgrade node phase --help" command.
+	// The flag "--skip-phases" takes precedence over this field.
+	SkipPhases []string
```
forgot to add - i don't see a field corresponding to
--ignore-preflight-errors

chendave replied:
> IMO, we should drop the comment about precedence.

done

chendave commented:
> one Q we have to ask is do we want to move diff specific options in a substruct?

I thought about this, it's would be more neat to have a substruct, but we need to introduce several new substruct to hold the config for different commands.

I am working on respin a new PR, let's see how many config item left in the next version, I am open to this, also, @SataQiu @pacoxu love to hear from you as well.

chendave commented:
thanks @neolit123 for the review, really great suggestion! I need to take several days to think about it and prepare a new PR.

chendave committed:
kubeadm: Adapt the `apply` and `plan` for `UpgradeConfiguration`

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave committed:
kubeadm: Adapt the `upgrade diff` for `UpgradeConfiguration`

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave committed:
kubeadm: Adapt the `upgrade node` for `UpgradeConfiguration`

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented:
/test pull-kubernetes-node-e2e-containerd

chendave commented:
/test pull-kubernetes-node-e2e-containerd    (unrelated flaky)

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	DryRun bool
+
+	// NodeRegistration holds the node specific information needed for upgrade, e.g. CRISocket, kubeadm needs this info in the phase of preflight to pull the required images
+	NodeRegistration NodeRegistrationOptions
```
Having a `initConfiguration` to hold those config fetched from cluster, and removed all those from `upgradeConfiguration`.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+	// In HA setups, this differs from ClusterConfiguration.ControlPlaneEndpoint in the sense that ControlPlaneEndpoint
+	// is the global endpoint for the cluster, which then loadbalances the requests to each individual API server.
+	// kubeadm will try to auto-detect the IP of the default interface and use that.
+	LocalAPIEndpoint APIEndpoint
```
removed as well.

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -503,5 +503,75 @@ type ResetConfiguration struct {
+
+	// SchedulerManifestPath is the path to API server manifest
+	// Default: /etc/kubernetes/manifests/kube-scheduler.yaml
+	SchedulerManifestPath string
```
removed.

chendave commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -91,39 +91,38 @@ func loadConfig(cfgPath string, client clientset.Interface, skipComponentConfigs
 		// Now default and validate the configs
-		componentconfigs.Default(&initCfg.ClusterConfiguration, &initCfg.LocalAPIEndpoint, &initCfg.NodeRegistration)
-		if errs := componentconfigs.Validate(&initCfg.ClusterConfiguration); len(errs) != 0 {
+		componentconfigs.Default(&cfg.ClusterConfiguration, &cfg.LocalAPIEndpoint, &cfg.NodeRegistration)
```
ditto, not a config item in `upgradeCfg` now.

chendave commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -133,7 +132,7 @@ func enforceRequirements(flags *applyPlanFlags, args []string, dryRun bool, upgr
 
 	var newK8sVersion string
-	cfg, legacyReconfigure, err := loadConfig(flags.cfgPath, client, !upgradeApply, printer)
+	cfg, reconfigure, err := loadConfig(flags.cfgPath, client, !upgradeApply, printer)
```
there is no reconfiguration anymore, the `clusterConfiguration`, `initConfiguration` and componet config are now deprecated and ignored.

chendave commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -179,7 +178,7 @@ func enforceRequirements(flags *applyPlanFlags, args []string, dryRun bool, upgr
 	}
-	// Also set the union of pre-flight errors to InitConfiguration, to provide a consistent view of the runtime configuration:
+	// Also set the union of pre-flight errors to UpgradeConfiguration, to provide a consistent view of the runtime configuration:
 	cfg.NodeRegistration.IgnorePreflightErrors = sets.List(ignorePreflightErrorsSet)
```
Done

chendave commented on cmd/kubeadm/app/cmd/upgrade/common_test.go:
```diff
@@ -261,3 +263,96 @@ func TestIsKubeadmConfigPresent(t *testing.T) {
 	}
 }
+
+func TestLoadUpgradeConfigurationFromFile(t *testing.T) {
```
removed.

chendave commented on cmd/kubeadm/app/util/config/upgradeconfiguration.go:
```diff
@@ -0,0 +1,147 @@
+/*
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/validation/validation_test.go:
```diff
@@ -717,6 +720,8 @@ func TestValidateMixedArguments(t *testing.T) {
 		f.String("foo", "", "flag bound to config object")
 		f.StringSliceVar(&ignorePreflightErrors, "ignore-preflight-errors", ignorePreflightErrors, "flag not bound to config object")
+		f.Bool("allow-experimental-upgrades", true, "upgrade flags for plan and apply command")
+		f.StringVar(&manifest, "api-server-manifest", manifest, "upgrade flags for diff command")
```
will deprecate them after the upgrade config API is ready.

chendave commented on cmd/kubeadm/app/componentconfigs/kubelet.go:
```diff
@@ -69,7 +69,7 @@ var kubeletHandler = handler{
 }
 
-func kubeletConfigFromCluster(h *handler, clientset clientset.Interface, clusterCfg *kubeadmapi.ClusterConfiguration) (kubeadmapi.ComponentConfig, error) {
+func kubeletConfigFromCluster(h *handler, clientset clientset.Interface, _ *kubeadmapi.ClusterConfiguration) (kubeadmapi.ComponentConfig, error) {
```
all cleanup has been split out.

chendave commented on cmd/kubeadm/app/phases/upgrade/postupgrade.go:
```diff
@@ -297,26 +297,3 @@ func GetKubeletDir(dryRun bool) (string, error) {
 	return kubeadmconstants.KubeletRunDirectory, nil
 }
-
-// moveFiles moves files from one directory to another.
```
Done

chendave commented:
update 8/28/23 (diff with last reversion)
------------------------------------------------
1. removed all `ManifestPath` from upgrade config API, we can deprecate those flags after this one is merged.
reason for that is: this path is hardcoded for `init` and `join`, it's only potentially used for `dry-run` case where the path could be redefined, config them are over-thinking.

2. Instead of having some immutable config item, e.g. `NodeRegistration` in the upgradeConfiguration API, keeping to use initConfiguration to hold the clusterCfg and other immutable configuration fetched from cluster, and keep the upgradeConfiguration to focus on the upgrade specific config items (the logic is similar with `kubeadm join`).

3. deprecate the `clusterConfiguration`, `initConfiguration` and componet config for `upgrade plan` and `upgrade apply`, use can still config them in a config file but they will be ignored.

4. config `ignore-preflight-errors`, `FeatureGates` is added to the `upgradeConfiguration`.

5. split all cleanup commit out for separate review.

6. try to validate the pr with the following command, 
`kubeadm join --dry-run`  - success
`kubeadm -- upgrade plan` - success
`kubeadm upgrade apply v1.28.1` - success
> [upgrade/successful] SUCCESS! Your cluster was upgraded to "v1.28.1". Enjoy!

7. other general fixup.


need feedback
----------------
1. flat struct or nest struct
> one Q we have to ask is do we want to move diff specific options in a substruct?

> e.g. UpgradeConfiguration.Diff.SomeOptionSpecificToDiff

I am open to either of them, the flag config will be easier and unrelated config will be ignored for a specific command.

e.g. `kubeadm upgrade diff` will dismiss all config but `DiffContextLines`

I am happy to change the struct if anyone of you have any preference.

2. Do we need to support component config configuration in a config file?
> should we allow component configs 

I incline to say no,  and actually they are ignored if them are configured in the current impl, the reason for that is, 
- use can edit the configmap for customization. 
`kubeadm upgrade node phase control-plane` need to update the configmap `kubeadm-config` to set the `kubernetesVersion`, user can do the something for component configs. 
- the logic will be much simpler and easy to maintain.

If you agree with that, I can refactor the logic for `kubeadm upgrade plan` to get rid of component config from config API where it reads the version of components from the config file.

see the code from here https://github.com/kubernetes/kubernetes/blob/91344b4008f6c5dfce82541b0c96cb8670b8350c/cmd/kubeadm/app/cmd/upgrade/plan.go#L271

chendave commented:
@neolit123 , I hope I haven't missed any comment from your last review, but if there is pls just let me know.

neolit123 commented:
> ## update 8/28/23 (diff with last reversion)
> 1. removed all `ManifestPath` from upgrade config API, we can deprecate those flags after this one is merged.
>    reason for that is: this path is hardcoded for `init` and `join`, it's only potentially used for `dry-run` case where the path could be redefined, config them are over-thinking.

ack

> 2. Instead of having some immutable config item, e.g. `NodeRegistration` in the upgradeConfiguration API, keeping to use initConfiguration to hold the clusterCfg and other immutable configuration fetched from cluster, and keep the upgradeConfiguration to focus on the upgrade specific config items (the logic is similar with `kubeadm join`).

ok, this seems better.

> 3. deprecate the `clusterConfiguration`, `initConfiguration` and componet config for `upgrade plan` and `upgrade apply`, use can still config them in a config file but they will be ignored.

what about "upgrade node"?
generally --config for upgrade* should only accept the new UpgradeConfiguration kind and error for anything else.
we could make this a v1beta4 only change - i.e. we could preserve the old behavior for v1beta3 kinds, but i think that's too complicated. 

> 4. config `ignore-preflight-errors`, `FeatureGates` is added to the `upgradeConfiguration`.

feature gates is a form of mutation. while we don't have a mutating way to change FGs on upgrade, following the no-mutation principle we should not have this? generally, open to ideas what to do with this one, given we already have a flag.

> 5. split all cleanup commit out for separate review.

thanks

> 6. try to validate the pr with the following command,
>    `kubeadm join --dry-run`  - success
>    `kubeadm -- upgrade plan` - success
>    `kubeadm upgrade apply v1.28.1` - success
> 
> > [upgrade/successful] SUCCESS! Your cluster was upgraded to "v1.28.1". Enjoy!
> 
> 7. other general fixup.
> 
> ## need feedback

this PR shows that we haven't done the design stage properly.
a google design doc should have been done initially where we decide what should be added and not.
only once that doc is approved we should have a PR.

> 1. flat struct or nest struct
> 
> > one Q we have to ask is do we want to move diff specific options in a substruct?
> 
> > e.g. UpgradeConfiguration.Diff.SomeOptionSpecificToDiff
> 
> I am open to either of them, the flag config will be easier and unrelated config will be ignored for a specific command.
> 
> e.g. `kubeadm upgrade diff` will dismiss all config but `DiffContextLines`
> 
> I am happy to change the struct if anyone of you have any preference.

having substructs seems better to me.
can we have such for plan, diff as well?

> 
> 2. Do we need to support component config configuration in a config file?
> 
> > should we allow component configs
> 
> I incline to say no, and actually they are ignored if them are configured in the current impl, the reason for that is,
> 
> * use can edit the configmap for customization.
>   `kubeadm upgrade node phase control-plane` need to update the configmap `kubeadm-config` to set the `kubernetesVersion`, user can do the something for component configs.
> * the logic will be much simpler and easy to maintain.
> 
> If you agree with that, I can refactor the logic for `kubeadm upgrade plan` to get rid of component config from other occurrences where it reads the version of components from the config file.
> 
> see the code from here
> 
> https://github.com/kubernetes/kubernetes/blob/91344b4008f6c5dfce82541b0c96cb8670b8350c/cmd/kubeadm/app/cmd/upgrade/plan.go#L271

i'm also voting none. if needed we can add a separate subcommand, perhaps.
the component config space has been quite problematic, generally.

neolit123 commented:
> @neolit123 , I hope I haven't missed any comment from your last review, but if there is pls just let me know.

i going to have to find time to have a look at the code changes later this week. perhaps on Wed.
thank you for the updates.

chendave commented:
> what about "upgrade node"?

same for `upgrade node`, they will be ignored as well.


> i.e. we could preserve the old behavior for v1beta3 kinds, but i think that's too complicated.

It's predictable that will be complicated to have v1beta3 to stick with the old behavior. I'm personally lean forward to not do as that.

> feature gates is a form of mutation. while we don't have a mutating way to change FGs on upgrade, following the no-mutation principle we should not have this?

`kubeadm upgrade plan` and `kubeadm upgrade apply` has the feature gate flags.

> kubeadm upgrade apply --help | grep gate
      --feature-gates string               A set of key=value pairs that describe feature gates for various features. Options are:


> having substructs seems better to me.
can we have such for plan, diff as well?

Okay, I will update in the next push along with other fix.

> i'm also voting no. if needed we can add a separate subcommand, perhaps.

Okay, will refactor in the next push.

> i going to have to find time to have a look at the code changes later this week. perhaps on Wed.

thanks @neolit123 !  I will push along with the fixing of those comments if there is any.

chendave commented:
> feature gates is a form of mutation. while we don't have a mutating way to change FGs on upgrade, following the no-mutation principle we should not have this?

> > kubeadm upgrade plan and kubeadm upgrade apply has the feature gate flags.

@neolit123 , I don't have enough context for why `kubeadm upgrade plan` and `kubeadm upgrade apply` has the flag to define the feature gate.

It's a little mess here, now the definition might come from `InitConfiguration` fetched from cluster, or it could be from the `UpgradeConfiguration` after this PR, or it could be from the flag by the CLI. 

I'd suggest to deprecate this flag for `kubeadm upgrade plan` and `kubeadm upgrade apply` if there is no special consideration, and stick to the FG fetched from cluster just like other commands, WDYT?

neolit123 commented:
> It's a little mess here, now the definition might come from `InitConfiguration` fetched from cluster, or it could be from the `UpgradeConfiguration` after this PR, or it could be from the flag by the CLI.
> 

i also don't recall the reasons. likely mutation or inspection - what output would it be with feature x turned on.
but parts of upgrade are definitely messy.

> I'd suggest to deprecate this flag for `kubeadm upgrade plan` and `kubeadm upgrade apply` if there is no special consideration, and stick to the FG fetched from cluster just like other commands, WDYT?

+1 to not add FGs to upgrade configuration and deprecate the flags. you can make the change in the pr  but this needs agreement with the rest of the maintainers..

it has implications around mutations - how can the user change a gate during upgrade. sometime this is required

maybe we can document it at k8s.io

chendave commented:
> +1 to not add FGs to upgrade configuration and deprecate the flags. you can make the change in the pr but this needs agreement with the rest of the maintainers..

@SataQiu @pacoxu , can you pls give some input for this? I will update this pr to not include the FG in the upgradeConfiguration, and a follow up pr to deprecate the flag if there is no objection.

chendave commented:
> If we remove the feature-gate flag in upgrade apply, users have to update the configmap to change a FG of kubeadm during upgradation.(Especially for some newly added FG in new versions: user may want to use a new alpha FG in the new version, for instance.)

I suggest to edit the configmap instead and document this, pls see the easier discussion which is related with this.

> should we allow component configs

>> I incline to say no, and actually they are ignored if them are configured in the current impl, the reason for that is,

>> use can edit the configmap for customization.
kubeadm upgrade node phase control-plane need to update the configmap kubeadm-config to set the kubernetesVersion, user can do the something for component configs.

this should be same with component config (we are not going to support upgrade with a config file with component config), and the only way to set the kubernetesVersion for `kubeadm upgrade node phase control-plane` is edit configmap.

chendave commented:
> upgrade node obviously should not have --feature-gates flag, as it will make nodes not use the same configuration.

yes, only `plan` and `apply` has the flag.

chendave commented:
I will update in the next few days.

SataQiu commented:
>  +1 to not add FGs to upgrade configuration and deprecate the flags. 

Emm... I'm +0 for this. Although modifying the configmap can be a workaround, it adds more complexity to achieve automation(for example, some tools based on kubeadm), and it may make it difficult for users to use some new features in the feature. Users have to understand the internal mechanism of kubeadm, which seems to degrade kubeadm UX.
Of course, it seems acceptable if we have good documentation about it.

neolit123 commented:
> > +1 to not add FGs to upgrade configuration and deprecate the flags.
> 
> Emm... I'm +0 for this. Although modifying the configmap can be a workaround, it adds more complexity to achieve automation(for example, some tools based on kubeadm), and it may make it difficult for users to use some new features in the future. Users have to understand the internal mechanism of kubeadm, which seems to degrade kubeadm UX. Of course, it seems acceptable if we have good documentation about it.

one rule that we wanted to follow was to not allow the user to mutate the cluster on upgrade by passing --config. --feature-gates is similar in a way, but yes not having it degrades the current UX. what kubeadm has been missing is a way to "change a running cluster", that was under discussion to be added in the kubeadm operator or a new kubeadm subcommand, but it never happened.

i think if the users have to do it manually for now - docs can be added here:
https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-reconfigure/

chendave commented:
> but yes not having it degrades the current UX. what kubeadm has been missing is a way to "change a running cluster",

How about this? In order to not degrade the UX, I can leave it as is, feature gate can still be supported by the flags (without deprecation), but it won't be configurable through the config file.

neolit123 commented:
> > but yes not having it degrades the current UX. what kubeadm has been missing is a way to "change a running cluster",
> 
> How about this? In order to not degrade the UX, we can leave it as is, feature gate can still be supported by the flags (without deprecation), but it won't be configurable through the config file.

i am still +1 to deprecate the flags and document manual CM edits:
 https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-reconfigure/

@pacoxu what is your vote?

pacoxu commented:
deprecating it seems to be the correct path.

I still want to improve the “feature gate” update process UX. Probably we can find another way.

pacoxu commented:
https://github.com/kubernetes/kubeadm/blob/309309e5d2b2530ce80e3fe13b90189d85ed0ca4/kinder/pkg/cluster/manager/actions/actions.go#L51-L53

Correct me if I understand wrongly. In kubeadm CI, we use the `kubeadm upgrade apply --feature-gates= EtcdLearnerMode=true v1.x`.

neolit123 commented:
> https://github.com/kubernetes/kubeadm/blob/309309e5d2b2530ce80e3fe13b90189d85ed0ca4/kinder/pkg/cluster/manager/actions/actions.go#L51-L53
> 
> Correct me if I understand wrongly. In kubeadm CI, we use the `kubeadm upgrade apply --feature-gates= EtcdLearnerMode=true v1.x`.

yes, i think so.
since we are switching to manual (slightly degrading the ux, as we discussed) it means we have to patch the fg with bash in kinder once the flags are removed. i think we already had one job that does this.

given the cli is ga, the flags will take 1 year for removal....
the etcd fg can graduate before that.

it can be problem in the future...
but we can also design a mutation sub command for the cluster configuration. it needs a kep.

chendave commented:
update 09/08/23 (primary change)
--------------------------------------------

1. Remove the feature gate from the upgrade config
Based on the discussion above, we will consider to deprecate the flag of feature gate but it's not necessary in this PR.
i.e.  feature gate flags is still supported by the CLI flags at the moment, and it will merge with the feature gates from init config fetched from cluster (PR to replace instead of merge is on the fly).

2. substructure of each of the commands instead of a top level flatten config
The benefits for this change for my understanding is that you can provide one config file for all four upgrade commands instead of 4 copy of config file for each of the command.

3. Add one commit to remove the support of loading of the component configs from a config file, this impacts the `kubeadm upgrade plan`.
Along with the removal, the logic to parse the config file to decide whether a manual upgrade for the component configs is needed is removed as well.

chendave commented:
/test pull-kubernetes-e2e-kind-ipv6

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -511,3 +511,105 @@ type Arg struct {
+	metav1.TypeMeta
+
+	// Apply holds a list of elements that is configurable for "kubeadm upgrade apply" command.
+	Apply UpgradeApplyConfiguration
```
these should be pointers.

chendave replied:
Actually, I have a version to make this a pointer before pushing this update.

I am in favor of non-pointers here, because, 
- pointer here will make the access to the field of the substructure more complicit to handle with, I need make sure this is not a nil pointer and then make sure its field is not a nil pointer as well. e.g. 
If I want to access `UpgradeApplyConfiguration.AllowRCUpgrades`, some check is needed to make sure both `UpgradeApplyConfiguration` and `AllowRCUpgrades` is not nil. The same check is applicable for all those fields.
Some tweak around `ValueFromFlagsOrConfig` has already done [here](https://github.com/kubernetes/kubernetes/pull/114062/files#diff-88bb647764f810575a6c3abf75e2986650cc9b60dbe1de6b7d31fba02106d060R153-R156)


- Almost all the substructure in `types.go` is not defined as pointer, e.g. `InitConfiguration.NodeRegistrationOptions` and `InitConfiguration.APIEndpoint`. I understand it might better for a pointer field if we access it directly to judge whether it's set or not. But I don't think we have too many cases to read it directly, this is somehow different with other components in the Kubernetes where the config in the config file is preferred instead of flags.

neolit123 replied:
there is like a tradeoff for both options. one + for pointers is validation. if both substructure A and B are not nil, we can throw an error e.g. both apply and plan. otherwise we have to a compare struct to a default empty struft which is not really possible for some types fileds.

neolit123 replied:
also we are talking about separate code paths if we are having the substructures. e.g. all apply code must use the apply configuration and we will know it's not nil.

we can also pass the apply configuration instead of the whole upgrade configuration but i can see that might be more difficult as upgrade is a bit of a mess.

chendave replied:
> if both substructure A and B are not nil, we can throw an error e.g. both apply and plan.

well, I don't think we should throw an error for this case, it's could be a config with all those config set, each specific command just need to care the config which is related.
One config file to hold all the config for all the commands is the main advantage compared with top-level flatten configuration.  So that you can just provide one config file instead of 4 config file.

neolit123 replied:
sgtm, but we need consensus with the others.
i find it irrigating that this was not discussed on a kep or a google doc.

appreciate your efforts, though.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -506,3 +506,130 @@ type Arg struct {
+
+	// Apply holds a list of elements that is configurable for "kubeadm upgrade apply" command.
+	// +optional
+	Apply UpgradeApplyConfiguration `json:"apply,omitempty"`
```
pointers.

chendave replied:
explained here: https://github.com/kubernetes/kubernetes/pull/114062#discussion_r1320522076

neolit123 commented:
> ## update 09/08/23 (primary change)
> 1. Remove the feature gate from the upgrade config
>    Based on the discussion above, we will consider to deprecate the flag of feature gate but it's not necessary in this PR.
>    i.e.  feature gate flags is still supported by the CLI flags at the moment, and it will merge with the feature gates from init config fetched from cluster (PR to replace instead of merge is on the fly).

+1 to deprecate in separate PRs and add the appropriate release notes.

> 2. substructure of each of the commands instead of a top level flatten config
>    The benefits for this change for my understanding is that you can provide one config file for all four upgrade commands instead of 4 copy of config file for each of the command.

had a quick look and i think i like it this way more.
others need to TAL as well.
it's a pity we did not discuss all of that on a google doc / KEP....

> 3. Add one commit to remove the support of loading of the component configs from a config file, this impacts the `kubeadm upgrade plan`.
>    Along with the removal, the logic to parse the config file to decide whether a manual upgrade for the component configs is needed is removed as well.

neolit123 commented:
mwe might be blocked merging PRs (especially upgrade) because of:
https://github.com/kubernetes/kubeadm/issues/2927

chendave commented:
@SataQiu @pacoxu  pls take a look when you are free, pls let me know is there any major concerns so I can update it accordingly.

chendave commented:
/hold

chendave commented:
Just some cleanup, no major update, pls see [diff](https://github.com/kubernetes/kubernetes/compare/6da36969367cffda08877aa16906707ac5864b29..5f9918ca76f24e91d27cd7f9112451c8fbe16844)

with the merge of this one: https://github.com/kubernetes/kubernetes/pull/120788 and also some additional warning when load the upgrade config file, see [warning](https://github.com/kubernetes/kubernetes/pull/114062/commits/a0bed08b5313a47930cea8c3c9c79736375c433b#diff-49f209c88d1ecf8c77c818d92db6c2411ea3638c969cc5baa10bc8a990e3e115R160-R165) here, the logic is much clearer now.

chendave commented:
@neolit123 @SataQiu @pacoxu pls let me know is there any further comments for this thanks!

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/apply.go:
```diff
@@ -100,39 +105,63 @@ func newCmdApply(apf *applyPlanFlags) *cobra.Command {
+	if ok {
+		flags.allowRCUpgrades = *allowRCUpgrades
+	} else {
+		return errors.New("type mismatch, AllowRCUpgrades is expected to be a pointer to bool")
```
a helper function can be used, maybe in common.go, so that the same error string is not copied many times?

```
getBoolPtrError := func(opt string) error {
	return errors.Errorf("type mismatch, %s is expected to be a pointer to bool", opt)	
}

```

chendave replied:
Done

neolit123 commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -59,81 +60,51 @@ func isKubeadmConfigPresent(docmap kubeadmapi.DocumentMap) bool {
+	// Also set the union of pre-flight errors to UpgradeConfiguration, to provide a consistent view of the runtime configuration.
+	// .Plan.IgnorePreflightErrors is not set as it's not used.
+	if upgradeApply {
+		upgradeCfg.Apply.IgnorePreflightErrors = sets.List(ignorePreflightErrorsSet)
```
is this causing a problem otherwise?
can we just check if `len(upgradeCfg.Apply.IgnorePreflightErrors) > 0` instead of defaulting it with a `sets.List()` here?

chendave replied:
I am not sure understanding it correctly, the line here is assigning value to (convert sets.Set[string] to []string) to `upgradeCfg.Apply.IgnorePreflightErrors`, not sure what the check you meant here, correct me pls.

neolit123 commented on cmd/kubeadm/app/util/marshal_test.go:
```diff
@@ -431,3 +431,37 @@ func TestGroupVersionKindsHasResetConfiguration(t *testing.T) {
+		expected bool
+	}{
+		{
+			name: "NoUpgradeConfiguration",
```
```suggestion
			name: "no UpgradeConfiguration found",
```

chendave replied:
Done

neolit123 commented on cmd/kubeadm/app/util/marshal_test.go:
```diff
@@ -431,3 +431,37 @@ func TestGroupVersionKindsHasResetConfiguration(t *testing.T) {
+			expected: false,
+		},
+		{
+			name: "UpgradeConfigurationFound",
```
```suggestion
			name: "UpgradeConfiguration is found",
```

chendave replied:
Done

neolit123 reviewed (commented):
thanks for the updates.
did another review pass. added some comments mostly around godoc / typos. minor suggestions here and there. still needs >= 2 reviewers.

the commit organization SGTM. the apply.go, node.go, diff.go...etc changes are quite difficult to read, but that code needs a big refactor and only then diffs will be readable.
also "upgrade apply" needs the phases, similarly to "upgrade node" that already has them. there is a k/kubeamd issue for that somewhere.

chendave commented:
thanks @neolit123 for the review!

> but that code needs a big refactor and only then diffs will be readable.

yes, let me see if I can split the this PR into more small commits, or I can do that in a separate PRs (problem is that most of refactor there cannot be moved out without the context of the upgrade API).


> also "upgrade apply" needs the phases

do you mean https://github.com/kubernetes/kubeadm/issues/1318 and https://github.com/kubernetes/kubernetes/pull/108767, maybe we can address that separately.

 
> still needs >= 2 reviewers.

@SataQiu @pacoxu heads up pls, most important thing needs to sync with you is: https://github.com/kubernetes/kubernetes/pull/114062#discussion_r1320528945, pls let me know is there any objection from you, thanks!

chendave commented:
just rebase, nothing more was changed, will try to split this pr to more small commits for easier review later.

chendave commented:
/test pull-kubernetes-e2e-kind-ipv6

chendave committed:
kubeadm: refactor `enforceRequirements` to adapt with `UpgradeConfiguration`

Instead of load from config file to populate `InitConfiguration`,
the config file is expected to hold the config for upgrade only.

The old logic is that if the config file has "initConfiguration/
ClusterConfiguration" defined, then it assumes that user wants to
reconfigure the cluster based on the new configuration.

upgrade itself should be non-mutable, so the logic to `reconfigure`
will not be supported, and the init config should be fetched from
cluster instead.

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+}
+
+// UpgradeNodeConfiguration contains a list of configurable elements which is specific for "kubeadm upgrade node" command.
+type UpgradeNodeConfiguration struct {
```
update the https://github.com/kubernetes/kubeadm/issues/2890 to have a TODO item there.

chendave commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go:
```diff
@@ -224,3 +225,22 @@ func SetDefaults_EnvVar(obj *EnvVar) {
 }
+
+// SetDefaults_UpgradeConfiguration assigns default values for the UpgradeConfiguration
+func SetDefaults_UpgradeConfiguration(obj *UpgradeConfiguration) {
```
Done

> there is a new codegen directive called +default, that can be used for both applying the default magically without a defaulting function written manually, and also being useful for users reading the API.

Good to know, we can hold until the feature is stable.

chendave committed:
kubeadm: remove function `loadConfig` from code base

The function use to load "InitConfiguration/ClusterConfiguration/
componentConfig" from a config file, now that we have a `upgradeConfiguration`
API, the function `loadConfig` can be removed now.

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+// UpgradeApplyConfiguration contains a list of configurable elements which is specific for  "kubeadm upgrade apply" command.
+type UpgradeApplyConfiguration struct {
+	// AllowExperimentalUpgrades instructs kubeadm to show unstable versions of Kubernetes as an upgrade
+	// alternative and allow upgrading to an alpha/beta/release candidate version of Kubernetes.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+	AllowExperimentalUpgrades *bool
+
+	// Enable AllowRCUpgrades will show release candidate versions of Kubernetes as an upgrade alternative and
+	// allow upgrading to a release candidate version of Kubernetes.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
 	EncryptionAlgorithmRSA EncryptionAlgorithmType = "RSA"
 )
+
+// UpgradeApplyConfiguration contains a list of configurable elements which is specific for  "kubeadm upgrade apply" command.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+	DiffContextLines int
+}
+
+// UpgradeNodeConfiguration contains a list of configurable elements which is specific for "kubeadm upgrade node" command.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+	Patches *Patches
+}
+
+// UpgradePlanConfiguration contains a list of configurable elements which is specific for "kubeadm upgrade plan" command.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+	PrintConfig *bool
+}
+
+// UpgradeDiffConfiguration contains a list of configurable elements which is specific for "kubeadm upgrade diff" command.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// UpgradeConfiguration contains a list of elements that is specific "kubeadm upgrade"-only runtime information.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+	// Diff holds a list of elements that is configurable for "kubeadm upgrade diff" command.
+	Diff UpgradeDiffConfiguration
+
+	// Node holds a list of elements that is configurable for "kubeadm upgrade node" command.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -535,3 +535,105 @@ const (
+	// Node holds a list of elements that is configurable for "kubeadm upgrade node" command.
+	Node UpgradeNodeConfiguration
+
+	// Plan holds a list of elements that is configurable for "kubeadm upgrade plan" command.
```
Done

chendave commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -35,6 +35,9 @@ limitations under the License.
 //     "ECDSA" (uses the P-256 eliptic curve).
+//   - The UpgradeConfiguration API type is now supported in v1beta4. Users are able to upgrade a cluster or node by passing
+//     a --config file to each sub-command of "kubeadm upgrade". Component Configs, InitConfiguration, ClusterConfiguration are
+//     deprecated and will be ignored in the upgrade config file.
```
Done

chendave commented:
update 10/20/23
===========

- address @neolit123 's comments
- added two separate commits for refactor https://github.com/kubernetes/kubernetes/pull/114062/commits/87770596ce5bb7c40ccbb84247d200f450d46fb5 and https://github.com/kubernetes/kubernetes/pull/114062/commits/65f515232a27541b5c133827014e1048d33d4a4d
- re-org each commits and cleanup all unnecessary changes.

Overall, the big change for this update is refactoring, I did my best to make those changes easier to review.

@neolit123 @pacoxu @SataQiu  PTAL when you got a chance, thanks!

chendave commented:
/retest-required

chendave commented on cmd/kubeadm/app/cmd/upgrade/common.go:
```diff
@@ -105,50 +141,40 @@ func enforceRequirements(flags *applyPlanFlags, args []string, dryRun bool, upgr
-	klog.V(1).Info("running preflight checks")
-	if err := runPreflightChecks(client, ignorePreflightErrorsSet, printer); err != nil {
-		return nil, nil, nil, err
-	}
```
this diff is necessary, I must move this code block up before the call to `FetchInitConfigurationFromCluster`, so that the interaction with cluster will not happen, this is because some testcases want to check the logic around this `runPreflightChecks`.

see this testcase:
cmd/kubeadm/app/cmd/upgrade TestEnforceRequirements/Fail_pre-flight_check

and the log from here: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/114062/pull-kubernetes-unit/1715315999886020608/build-log.txt

chendave committed:
kubeadm: define more constant for `kubeadm upgrade`

Those constants will be used for `kubeadm upgrade plan` or `kubeadm upgrade apply`

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented:
Just note I don't have time on this at the moment, someone else can help to continue the efforts as it's pretty close I believe, or else I can come back to this when some mess from me are all settled.

calvin0327 commented:
@chendave I'm interested in this, can I work on this?

neolit123 commented:
closing in favor of https://github.com/kubernetes/kubernetes/pull/123068
thanks for the work on this!

neolit123 commented:
/close
